File: app/src/main/java/com/elegion/android/ui/login/LoginFragment.java
package com.elegion.android.ui.login;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;
import com.elegion.android.R;
import com.elegion.android.data.Repository;
import com.elegion.android.ui.base.fragment.BaseNoInternetFragment;

import butterknife.BindView;
import butterknife.OnClick;
import butterknife.OnTextChanged;

/**
 * @author Mike
 */

public class LoginFragment extends BaseNoInternetFragment implements LoginView {
    @BindView(R.id.top_layout)
    View mTopLayout;

    @InjectPresenter
    LoginPresenter mPresenter;

    @ProvidePresenter
    LoginPresenter providePresenter() {
        return new LoginPresenter(Repository.get(getActivity()));
    }

    public static Fragment newInstance() {
        return new LoginFragment();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fr_login, container);
    }

    @OnClick(R.id.login_btn)
    public void onLoginClick() {
        mPresenter.login();
    }

    @OnTextChanged(R.id.email)
    public void onEmailChanged(CharSequence text) {
        mPresenter.setEmail(text.toString());
    }

    @OnTextChanged(R.id.password)
    public void onPasswordChanged(CharSequence text) {
        mPresenter.setPassword(text.toString());
    }

    @Override
    protected View[] getViews() {
        return new View[] {mTopLayout};
    }

    @Override
    public void tryAgain() {
        mPresenter.login();
    }

    @Override
    public void loginSuccessful() {

    }
}

File: app/src/main/kotlin/com/elegion/android/template/ui/login/LoginFragment.kt
package com.elegion.android.template.ui.login

import android.text.TextWatcher
import android.view.View
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.ProvidePresenter
import com.elegion.android.template.R
import com.elegion.android.template.data.Repository
import com.elegion.android.template.extension.android.widget.addTextChangedListener
import com.elegion.android.template.ui.base.fragment.BaseNoInternetFragment
import com.elegion.android.template.ui.features.FeaturesActivity
import kotlinx.android.synthetic.main.fr_login.*

class LoginFragment : BaseNoInternetFragment(), LoginView {
    @InjectPresenter
    internal lateinit var presenter: LoginPresenter

    private lateinit var loginEmailTexTWatcher: TextWatcher
    private lateinit var loginPasswordTexTWatcher: TextWatcher

    @ProvidePresenter
    internal fun providePresenter(): LoginPresenter = LoginPresenter(Repository.get(activity!!))

    override fun getLayout(): Int = R.layout.fr_login

    override fun onResume() {
        super.onResume()
        loginBtn.setOnClickListener { presenter.login() }
        loginLetMeInBtn.setOnClickListener { presenter.letMeIn() }
        loginEmailTexTWatcher = loginEmail.addTextChangedListener {
            onTextChanged { text, _, _, _ -> presenter.setEmail(text.toString()) }
        }
        loginPasswordTexTWatcher = loginPassword.addTextChangedListener {
            onTextChanged { text, _, _, _ -> presenter.setPassword(text.toString()) }
        }
    }

    override fun onPause() {
        super.onPause()
        loginBtn.setOnClickListener(null)
        loginEmail.removeTextChangedListener(loginEmailTexTWatcher)
        loginPassword.removeTextChangedListener(loginPasswordTexTWatcher)
    }

    override fun getViews(): Array<View> = arrayOf(loginTopLayout)

    override fun tryAgain() = presenter.login()

    override fun loginSuccessful() {
        startActivity(FeaturesActivity.makeIntent(activity!!))
        activity?.supportFinishAfterTransition()
    }

    companion object {
        fun newInstance() = LoginFragment()
    }
}
